// awilix-express/lib/index.mjs
import { asFunction, asClass } from 'awilix';
import { isClass } from 'awilix/lib/utils';
import assert from 'assert';
import { getStateAndTarget, findControllers, rollUpState, HttpVerbs } from 'awilix-router-core';
export * from 'awilix-router-core';
import { Router } from 'express';

/**
 * Creates either a function invoker or a class invoker, based on whether
 * the argument can be classified as a class or not. Uses Awilix' `isClass` utility.
 *
 * @param functionOrClass
 * The function or class to invoke.
 *
 * @param opts
 * Resolver options for the class/function.
 */
function makeInvoker(functionOrClass, opts) {
    return isClass(functionOrClass)
        ? /*tslint:disable-next-line*/
            makeClassInvoker(functionOrClass, opts)
        : /*tslint:disable-next-line*/
            makeFunctionInvoker(functionOrClass, opts);
}
/**
 * Returns a function that when called with a name,
 * returns another function to be used as Express middleware.
 * That function will run `fn` with the container cradle as the
 * only parameter, and then call the `methodToInvoke` on
 * the result.
 *
 * @param, {Function} fn
 * @return {(methodToInvoke: string) => (req) => void}
 */
function makeFunctionInvoker(fn, opts) {
    return makeResolverInvoker(asFunction(fn, opts));
}
/**
 * Same as `makeInvoker` but for classes.
 *
 * @param  {Class} Class
 * @return {(methodToInvoke: string) => (req) => void}
 */
function makeClassInvoker(Class, opts) {
    return makeResolverInvoker(asClass(Class, opts));
}
/**
 * Returns a function that when called with a method name,
 * returns another function to be used as Express middleware.
 * That function will run `container.build(resolver)`, and
 * then call the method on the result, passing in the Express request
 * and rest of the parameters.
 *
 * @param, {Resolver} resolver
 * @return {(methodToInvoke: string) => (req) => void}
 */
function makeResolverInvoker(resolver) {
    /**
     * 2nd step is to create a method to invoke on the result
     * of the resolver.
     *
     * @param  methodToInvoke
     * @return {(req) => void}
     */
    return function makeMemberInvoker(methodToInvoke) {
        /**
         * The invoker middleware.
         *
         * @param  {E.Request} req
         * @param  {...*} rest
         * @return {*}
         */
        return function memberInvoker(req, ...rest) {
            const container = req.container;
            const resolved = container.build(resolver);
            assert(methodToInvoke, `methodToInvoke must be a valid method type, such as string, number or symbol, but was ${String(methodToInvoke)}`);
            if (!resolved[methodToInvoke]) {
                throw Error(`The method attempting to be invoked, '${String(methodToInvoke)}', does not exist on the controller`);
            }
            return asyncErrorWrapper(resolved[methodToInvoke].bind(resolved))(req, ...rest);
        };
    };
}
/**
 * Injects dependencies into the middleware factory when the middleware is invoked.
 *
 * @param factory
 */
function inject(factory) {
    const resolver = getResolver(factory);
    /**
     * The invoker middleware.
     */
    return function middlewareFactoryHandler(req, ...rest) {
        const container = req.container;
        const resolved = container.build(resolver);
        return asyncErrorWrapper(resolved)(req, ...rest);
    };
}
/**
 * Wraps or returns a resolver.
 */
function getResolver(arg) {
    if (typeof arg === 'function') {
        /*tslint:disable-next-line*/
        return asFunction(arg);
    }
    return arg;
}
/**
 * Wraps a middleware, detects if it returns a Promise and calls next() with the caught error, if necessary.
 * @param fn
 */
function asyncErrorWrapper(fn) {
    return function asyncWrappedMiddleware(req, res, next) {
        const returnValue = fn(req, res, next);
        if (returnValue &&
            returnValue.catch &&
            typeof returnValue.catch === 'function') {
            return returnValue.catch((err) => {
                next(err);
            });
        }
        else {
            return returnValue;
        }
    };
}

/**
 * Express middleware factory that will create and attach
 * a scope onto a content.
 *
 * @param  {AwilixContainer} container
 * @return {Function}
 */
function scopePerRequest(container) {
    return function scopePerRequestMiddleware(req, res, next) {
        req.container = container.createScope();
        return next();
    };
}

/**
 * Registers one or multiple decorated controller classes.
 *
 * @param ControllerClass One or multiple "controller" classes
 *        with decorators to register
 */
function controller(ControllerClass) {
    const router = Router();
    if (Array.isArray(ControllerClass)) {
        ControllerClass.forEach((c) => _registerController(router, getStateAndTarget(c)));
    }
    else {
        _registerController(router, getStateAndTarget(ControllerClass));
    }
    return router;
}
/**
 * Loads controllers for the given pattern.
 *
 * @param pattern
 * @param opts
 */
function loadControllers(pattern, opts) {
    const findOpts = { ...opts, absolute: true };
    const result = findControllers(pattern, findOpts);
    if (opts?.esModules) {
        return result.then((found) => {
            const router = Router();
            found.forEach(_registerController.bind(null, router));
            return router;
        });
    }
    const router = Router();
    result.forEach(_registerController.bind(null, router));
    return router;
}
/**
 * Reads the config state and registers the routes in the router.
 *
 * @param router
 * @param ControllerClass
 */
function _registerController(router, stateAndTarget) {
    if (!stateAndTarget) {
        return;
    }
    const { state, target } = stateAndTarget;
    const rolledUp = rollUpState(state);
    rolledUp.forEach((methodCfg, methodName) => {
        methodCfg.verbs.forEach((httpVerb) => {
            let method = httpVerb.toLowerCase();
            if (httpVerb === HttpVerbs.ALL) {
                method = 'all';
            }
            router[method](methodCfg.paths, ...methodCfg.beforeMiddleware, 
            /*tslint:disable-next-line*/
            makeInvoker(target)(methodName), ...methodCfg.afterMiddleware);
        });
    });
}

export { controller, inject, loadControllers, makeClassInvoker, makeFunctionInvoker, makeInvoker, makeResolverInvoker, scopePerRequest };
