// awilix-router-core/lib/index.mjs
import glob from 'fast-glob';
import { pathToFileURL } from 'node:url';

/**
 * Http methods.
 */
const HttpVerbs = {
    GET: 'GET',
    HEAD: 'HEAD',
    POST: 'POST',
    PUT: 'PUT',
    DELETE: 'DELETE',
    CONNECT: 'CONNECT',
    OPTIONS: 'OPTIONS',
    PATCH: 'PATCH',
    ALL: '*',
};

/**
 * Asserts the condition and returns it, throws otherwise.
 *
 * @param condition Must be truthy, else throws. Returns the condition is truthy.
 * @param message The message to use when throwing.
 */
function invariant(condition, message) {
    if (!condition) {
        throw new Error(message);
    }
    return condition;
}

/**
 * Returns unique items.
 */
function uniq(src) {
    const result = [];
    src.forEach((t) => {
        if (result.indexOf(t) === -1) {
            result.push(t);
        }
    });
    return result;
}

/**
 * Symbol used for getting and setting the config state.
 */
const STATE = Symbol('Router Config');
/**
 * Symbol used to determine whether or not something is a controller builder.
 */
const IS_CONTROLLER_BUILDER = Symbol('Is Controller Builder');

/**
 * Rolls up state so paths are joined, middleware rolled into
 * the correct order, etc.
 *
 * @param state
 */
function rollUpState(state) {
    const result = new Map();
    state.methods.forEach((method, key) => {
        result.set(key, {
            paths: concatPaths(state.root.paths, method.paths),
            beforeMiddleware: [
                ...state.root.beforeMiddleware,
                ...method.beforeMiddleware,
            ],
            afterMiddleware: [
                ...method.afterMiddleware,
                ...state.root.afterMiddleware,
            ],
            verbs: method.verbs,
        });
    });
    return result;
}
/**
 * Given a decorated class or a controller builder, returns a normalized
 * target + state object. For example, if using the controller builder,
 * we need to use the `target` property. If using decorators, the required
 * value *is* the target.
 *
 * @param src
 * @returns The normalized target + state, or `null` if not applicable.
 */
function getStateAndTarget(src) {
    const state = getState(src);
    if (!state) {
        return null;
    }
    const target = src[IS_CONTROLLER_BUILDER]
        ? src.target
        : src;
    return { target, state };
}
/**
 * Adds a route to the state.
 *
 * @param state
 * @param methodName
 * @param path
 */
function addRoute(state, methodName, path) {
    const config = getOrCreateConfig(state, methodName);
    return updateConfig(state, methodName, {
        paths: uniq([...config.paths, path]),
    });
}
/**
 * Adds middleware that runs before the method on the specified config.
 *
 * @param state
 * @param methodName
 * @param middleware
 */
function addBeforeMiddleware(state, methodName, middleware) {
    const config = getOrCreateConfig(state, methodName);
    return updateConfig(state, methodName, {
        beforeMiddleware: addMiddleware(config.beforeMiddleware, middleware),
    });
}
/**
 * Adds middleware that runs after the method on the specified config.
 *
 * @param state
 * @param methodName
 * @param middleware
 */
function addAfterMiddleware(state, methodName, middleware) {
    const config = getOrCreateConfig(state, methodName);
    return updateConfig(state, methodName, {
        afterMiddleware: addMiddleware(config.afterMiddleware, middleware),
    });
}
/**
 * Adds middleware that runs after the method on the specified config.
 *
 * @param state
 * @param methodName
 * @param value
 */
function addHttpVerbs(state, methodName, value) {
    const config = getOrCreateConfig(state, methodName);
    return updateConfig(state, methodName, {
        verbs: uniq([...config.verbs, ...value]),
    });
}
/**
 * Gets or creates a method config.
 *
 * @param state
 * @param methodName
 */
function getOrCreateConfig(state, methodName) {
    const config = methodName === null ? state.root : state.methods.get(methodName);
    if (!config) {
        return createRouteConfig();
    }
    return config;
}
/**
 * Gets the config state from the target.
 *
 * @param target
 */
function getState(target) {
    return ((target && (target.prototype ? target.prototype[STATE] : target[STATE])) ||
        null);
}
/**
 * Sets the config state on the target.
 *
 * @param target
 * @param state
 */
function setState(target, state) {
    if (target.prototype) {
        target.prototype[STATE] = state;
    }
    else {
        target[STATE] = state;
    }
    return state;
}
/**
 * Updates the state on the specified target by invoking the callback with the previous state.
 *
 * @param target
 * @param updater
 */
function updateState(target, updater) {
    setState(target, updater(getOrInitStateForDecoratorTarget(target)));
}
/**
 * Gets or initializes the state for a decorated target
 *
 * @param target
 * @param name
 */
function getOrInitStateForDecoratorTarget(target) {
    return getState(target) || createState();
}
/**
 * Creates a new state object.
 */
function createState() {
    const state = {
        root: createRouteConfig(),
        methods: new Map(),
    };
    return state;
}
/**
 * Updates a config on a state, returns the new state.
 *
 * @param state
 * Existing state.
 *
 * @param methodName
 * If null, updates the root config. Else, the method config.
 *
 * @param newConfig
 * Config to shallow-merge in.
 */
function updateConfig(state, methodName, newConfig) {
    const existing = getOrCreateConfig(state, methodName);
    const mergedConfig = {
        ...existing,
        ...newConfig,
    };
    // Root update is simple.
    if (methodName === null) {
        return {
            ...state,
            root: mergedConfig,
        };
    }
    // Filters out the entry we're replacing.
    const filteredEntries = Array.from(state.methods.entries()).filter(([key]) => key !== methodName);
    return {
        ...state,
        methods: new Map([...filteredEntries, [methodName, mergedConfig]]),
    };
}
/**
 * Creates a new route config object.
 */
function createRouteConfig() {
    return {
        paths: [],
        beforeMiddleware: [],
        afterMiddleware: [],
        verbs: [],
    };
}
/**
 * Adds a middleware to the end of the target array.
 *
 * @param targetArray
 * @param value
 */
function addMiddleware(targetArray, value) {
    return Array.isArray(value)
        ? [...targetArray, ...value]
        : [...targetArray, value];
}
/**
 * Concatenates root and method paths so we have one for each combination.
 */
function concatPaths(rootPaths, methodPaths) {
    if (rootPaths.length === 0) {
        return [...methodPaths];
    }
    const result = [];
    rootPaths.forEach((rootPath) => {
        if (methodPaths.length === 0) {
            result.push(rootPath);
        }
        else {
            methodPaths.forEach((methodPath) => {
                result.push(rootPath + methodPath);
            });
        }
    });
    return result;
}

/**
 * Registers a path for this class method.
 * You can add more than one.
 *
 * @param path
 */
function route(path) {
    return function routerDecorator(target, name = null) {
        updateState(target, (state) => addRoute(state, name, path));
    };
}
/**
 * Adds one or more middleware before the method middleware.
 * Can be applied to classes and methods. Class-level `before` middleware
 * runs before method-level ones.
 *
 * @param middleware
 */
function before(middleware) {
    return function routerDecorator(target, name = null) {
        updateState(target, (state) => addBeforeMiddleware(state, name, middleware));
    };
}
/**
 * Adds one or more middleware after the method middleware.
 * Can be applied to classes and methods. Class-level `after` middleware
 * runs after method-level ones.
 *
 * @param middleware
 */
function after(middleware) {
    return function routerDecorator(target, name = null) {
        updateState(target, (state) => addAfterMiddleware(state, name, middleware));
    };
}
/**
 * Assigns one or more HTTP verbs to class method.
 *
 * @param httpVerbs
 */
function verbs(httpVerbs) {
    return function verbsDecorator(target, name = null) {
        invariant(name, 'The "verbs" decorator can only be used on class verbs.');
        updateState(target, (state) => addHttpVerbs(state, name, httpVerbs));
    };
}
/**
 * The same as `verbs([HttpVerbs.GET])`
 */
const GET = () => verbs([HttpVerbs.GET]);
/**
 * The same as `verbs([HttpVerbs.HEAD])`
 */
const HEAD = () => verbs([HttpVerbs.HEAD]);
/**
 * The same as `verbs([HttpVerbs.POST])`
 */
const POST = () => verbs([HttpVerbs.POST]);
/**
 * The same as `verbs([HttpVerbs.PUT])`
 */
const PUT = () => verbs([HttpVerbs.PUT]);
/**
 * The same as `verbs([HttpVerbs.DELETE])`
 */
const DELETE = () => verbs([HttpVerbs.DELETE]);
/**
 * The same as `verbs([HttpVerbs.CONNECT])`
 */
const CONNECT = () => verbs([HttpVerbs.CONNECT]);
/**
 * The same as `verbs([HttpVerbs.OPTIONS])`
 */
const OPTIONS = () => verbs([HttpVerbs.OPTIONS]);
/**
 * The same as `verbs([HttpVerbs.PATCH])`
 */
const PATCH = () => verbs([HttpVerbs.PATCH]);
/**
 * The same as `verbs([HttpVerbs.ALL])`
 */
const ALL = () => verbs([HttpVerbs.ALL]);

/**
 * Finds classes using the specified pattern and options.
 *
 * @param pattern Glob pattern
 * @param opts Glob options and ES modules dynamic import module loading
 */
function findControllers(pattern, opts) {
    const result = glob.sync(pattern, opts);
    /* istanbul ignore next: can't cover because we are replacing the importer in tests. */
    const importFn = opts?.import ?? ((path) => import(path));
    if (opts && opts.esModules === true) {
        return Promise.all(result.map((path) => importFn(pathToFileURL(path).toString()).then(extractStateAndTargetFromExports))).then((controllers) => controllers.reduce((acc, cur) => acc.concat(cur), []));
    }
    else {
        return result
            .map((path) => {
            const exports$1 = require(path);
            return extractStateAndTargetFromExports(exports$1);
        })
            .reduce((acc, cur) => acc.concat(cur), []);
    }
}
function extractStateAndTargetFromExports(exports$1) {
    const items = [];
    if (exports$1) {
        const stateAndTarget = getStateAndTarget(exports$1);
        if (stateAndTarget) {
            items.push(stateAndTarget);
            return items;
        }
        // loop through exports - this will cover named as well as a default export
        for (const key of Object.keys(exports$1)) {
            const stateAndTarget = getStateAndTarget(exports$1[key]);
            if (stateAndTarget) {
                items.push(stateAndTarget);
            }
        }
    }
    return items;
}

/**
 * Configures routing config for a class or function to be invoked by a router.
 *
 * @example
 *   const api = ({ todoService }) => ({
 *     find: (ctx) => { ... }
 *   })
 *
 *   export default createController(api)
 *     .prefix('/todos')
 *     .before(bodyParser())
 *     .get('/', 'find')
 *     .get('/:id', 'get')
 *     .post('/:id', 'create', {
 *       before: [authenticate()],
 *     })
 *     .patch('/:id', 'update', {
 *       before: [authenticate()]
 *     })
 */
function createController(ClassOrFunction) {
    return createControllerFromState(ClassOrFunction, createState());
}
/**
 * Creates a builder from existing state.
 * This is used internally, but exported for convenience.
 *
 * @param ClassOrFunction The target to invoke.
 * @param state Existing state to continue building on.
 */
function createControllerFromState(ClassOrFunction, state) {
    const builder = {
        [STATE]: state,
        [IS_CONTROLLER_BUILDER]: true,
        target: ClassOrFunction,
        get: createVerbFunction(HttpVerbs.GET),
        post: createVerbFunction(HttpVerbs.POST),
        put: createVerbFunction(HttpVerbs.PUT),
        patch: createVerbFunction(HttpVerbs.PATCH),
        delete: createVerbFunction(HttpVerbs.DELETE),
        head: createVerbFunction(HttpVerbs.HEAD),
        options: createVerbFunction(HttpVerbs.OPTIONS),
        connect: createVerbFunction(HttpVerbs.CONNECT),
        all: createVerbFunction(HttpVerbs.ALL),
        prefix(path) {
            return createControllerFromState(ClassOrFunction, addRoute(state, null, path));
        },
        before(middleware) {
            return createControllerFromState(ClassOrFunction, addBeforeMiddleware(state, null, middleware));
        },
        after(middleware) {
            return createControllerFromState(ClassOrFunction, addAfterMiddleware(state, null, middleware));
        },
        verbs(verbs, path, method, opts) {
            state = addRoute(state, method, path);
            state = addHttpVerbs(state, method, verbs);
            if (opts) {
                if (opts.before) {
                    state = addBeforeMiddleware(state, method, opts.before);
                }
                if (opts.after) {
                    state = addAfterMiddleware(state, method, opts.after);
                }
            }
            return createControllerFromState(ClassOrFunction, state);
        },
    };
    return builder;
    /**
     * Creates a preconfigured verb function.
     *
     * @param verb
     */
    function createVerbFunction(verb) {
        return function configureRoute(path, method, opts) {
            return builder.verbs([verb], path, method, opts);
        };
    }
}

export { ALL, CONNECT, DELETE, GET, HEAD, HttpVerbs, OPTIONS, PATCH, POST, PUT, addAfterMiddleware, addBeforeMiddleware, addHttpVerbs, addRoute, after, before, createController, createControllerFromState, createRouteConfig, createState, findControllers, getOrCreateConfig, getOrInitStateForDecoratorTarget, getState, getStateAndTarget, rollUpState, route, setState, updateConfig, updateState, verbs };
